cache normalize_path in req_uninstall and is_local

DefaultRyan opened the pull request:
In recent performance testing on Windows, I found a significant amount of time was spent calling `normalize_path` in a loop, frequently with the same arguments. In fact, it was so much time, that `normalize_path` dominated the running time of `from_dist` and `add` in req/req_uninstall.py.

It seems prudent to avoid caching `normalize_path` itself, as symlinks and directory structure can change while a Python process is running. This change introduces `normalize_path_cached`, leaving `normalize_path` alone but allowing `UninstallPathSet` to cache frequently used paths during its operation.

### Test results

Extracting a segment of pip's functional tests I ran `pip install --force-reinstall --no-deps .tox/.tmp/package/1/pip-23.1.dev0.tar.gz` on my Windows dev box

|    | Before | After |
| -- | -- | -- |
| Total time | 23.1 s | 21.8 s |
| install_given_reqs (_internal/req/\_\_init\_\_.py) | 3.69 s | 2.79 s |
| uninstall (_internal/req/req_install.py) | 1.28 s |  0.415 s |
| normalize_path (_internal/utils/misc.ph) | 0.883 s | 0.046 s |
| # of calls to normalize_path | 3993 | 187 |

uranusjr commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -312,6 +312,7 @@ def __init__(self, dist: BaseDistribution) -> None:
         self._pth: Dict[str, UninstallPthEntries] = {}
         self._dist = dist
         self._moved_paths = StashedUninstallPathSet()
+        normalize_path_cached.cache_clear()
```
It seems very weird to me instantiating a new object of this class has a global side effect. The cache should probably live on this object instead of a global function?

DefaultRyan replied:
> It seems very weird to me instantiating a new object of this class has a global side effect. The cache should probably live on this object instead of a global function?

My thinking was that this might be more generally useful, but you raise a good point. I think I'd be fine with that, but could you help me with the syntax on that? I only know how to apply `lru_cache` to a function, not locally. In that case, perhaps we should give `is_local` its own cache?

DefaultRyan replied:
> Since this function is used by three values in this class that we know upfront anyway, maybe it’d be a good idea to make those values `cached_property`?

I don't think that works. https://docs.python.org/3/faq/programming.html#faq-cache-method-calls says:
> The cached_property approach only works with methods that do not take any arguments.

DefaultRyan replied:
Got it figured out.

DefaultRyan commented:
This will fix #11889.

uranusjr commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -312,21 +313,25 @@ def __init__(self, dist: BaseDistribution) -> None:
         self._pth: Dict[str, UninstallPthEntries] = {}
         self._dist = dist
         self._moved_paths = StashedUninstallPathSet()
+        self._normalize_path_cached = functools.lru_cache(maxsize=256)(normalize_path)
```
The maxsize definition is unnecessary (impossible to hit).

uranusjr replied:
p.s. worth adding a comment here to explain why the wrapping is implemented.

DefaultRyan replied:
> The maxsize definition is unnecessary (impossible to hit).

Without the `maxsize` definition, it caused errors in the Python 3.7 tests. I'm not sure, but I'm guessing that the simpler syntax only became possible in 3.8 due to this note? https://docs.python.org/3/whatsnew/3.8.html#functools

uranusjr replied:
Arguments are not required inside parentheses, you know.

uranusjr commented:
Also need to fix the pre-commit errors. (Test failures are unrelated.)

DefaultRyan commented:
pre-commit.ci autofix

DefaultRyan commented:
Addressed feedback - the Windows tests failed for what seems like an unrelated infrastructure issue. However, the UI that allows me to rerun a failed job has vanished. Is there a way somebody here could rerun them?

DefaultRyan commented:
Friendly ping @uranusjr or any other maintainers.

DefaultRyan commented:
Friendly ping again to @uranusjr

uranusjr reviewed (approved):
I think the implementation makes sense. Less sure whether it’s a good idea to add this so I’ll keep it open for now… If no-one objects I think it can go into 23.1.

pfmoore commented:
If it gets merged, it will go into 23.1 (I have no particular view either way as I haven't been following this change). Looks like it still needs some work to rebase it on current main and fix the CI.

To be clear, it doesn't need to be on the 23.1 milestone, if it's not ready when the release is being cut, I'll simply drop it from the milestone. It's not something that would hold up the release.

DefaultRyan commented:
Thanks for re-reviewing, and it looks like @sbidoul successfully rebased the branch and PR. Thanks!

Any more blockers?
